packages/api-contract/src/Abi.ts

```diff
@@ -16,16 +17,41 @@ export default class ContractAbi implements InterfaceAbi {
 
   public readonly messages: AbiMessages = {};
 
-  public constructor (abi: ContractABI) {
-    validateAbi(abi);
+  public readonly version: AbiVersion = AbiVersion.v2;
 
-    this.abi = abi;
-    this.deploy = createMethod('deploy', abi.deploy);
+  public constructor (json: ContractABIV1 | ContractABIV2) {
+    let abi;
+    switch (json.version) {
+      case AbiVersion.v2:
```

Comment:
Like we actually do in the types constructors, I would probably guard against the constructor getting . to big here.

So rather split this with 2 static functions, both returning `[abi, deploy, messages]` and then . do something like . the following -

```js
if (json.version === AbiVersion.v2) {
  [this.abi, this.deploy, this.messages] = Abi.decodeV2(json);
} else {
  [this.abi, this.deploy, this.messages] = Abi.decodeV1(json);
}
```